### PR TITLE
tests: replica syncing

### PR DESCRIPTION
**Pull request by annieke**
**Description**
- add l2_dtl and replica images
- add replica sync tests for dummy tx and erc20 transfer

**Additional context**
This is currently not running on CI, see discussions in https://github.com/ethereum-optimism/optimism/pull/972

**Metadata**
- Fixes OP-773

### CODE REVIEW

File: ops/README.md [hunk 1]
```diff
@@ -30,10 +30,19 @@ docker-compose \
...
+    verifier=1 \
     --build --detach
 ```
 
+Do the same to run an L2 DTL and replica combo. Note that the L2 DTL is a requirement for the replica!
+```
+docker-compose up \
+    --scale l2_dtl=1 --scale replica=1 \
+    --build --detach
+```
```
Comment (annieke) [inline ops/README.md:43]: as of now this still doesn't work from scratch when sequencer is at block 0, since we set `highestSyncedL2BlockNumber` to be 1 instead of 0 in the dtl [here](https://github.com/ethereum-optimism/optimism/blob/245136f14afed2a731a5ca5f9b5b071a3131961b/packages/data-transport-layer/src/services/l2-ingestion/service.ts#L76-L79). this triggers a [startBlock > endBlock error](https://github.com/ethereum-optimism/optimism/blob/245136f14afed2a731a5ca5f9b5b071a3131961b/packages/data-transport-layer/src/services/l2-ingestion/service.ts#L140-L149) and prevents the l2_dtl from syncing from scratch, but works during the integration tests since we submit more transactions

Comment (gakonst) [reply ops/README.md:43]: Should we be setting it to 0 instead?

File: ops/docker-compose.yml [hunk 2]
```diff
@@ -140,23 +140,67 @@ services:
...
+      -  ./envs/dtl.env
+    environment:
+        URL: http://deployer:8081/addresses.json
+        DATA_TRANSPORT_LAYER__SYNC_FROM_L1: 'false'
+        DATA_TRANSPORT_LAYER__SYNC_FROM_L2: 'true'
+        DATA_TRANSPORT_LAYER_DEFAULT_BACKEND: 'l2'
+        DATA_TRANSPORT_LAYER__L2_RPC_ENDPOINT: http://l2geth:8545
+        DATA_TRANSPORT_LAYER__L2_CHAIN_ID: 420
+    ports:
+      - ${L2_DTL_PORT:-7879}:7878
```
Comment (gakonst) [inline ops/docker-compose.yml:179]: Do we need another DTL? I thought that @tynes made it so that l2geth nodes can query if they want to use L1 or L2 as the DTL backend.

File: integration-tests/sync-tests/sync-replica.spec.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      commandOptions: ['--scale', 'l2_dtl=1', '--scale', 'replica=1'],
+    })
+
+    // Wait for replica to be looping
+    let logs = await replica.logs()
+    while (!logs.out.includes('Starting Sequencer Loop')) {
+      console.log(logs)
+      await sleep(500)
+      logs = await replica.logs()
+    }
```
Comment (gakonst) [inline integration-tests/sync-tests/sync-replica.spec.ts:33]: Why does a Replica run the sequencer loop? Shouldn't it be running the verifier loop?

**Review by gakonst — CHANGES_REQUESTED** [review]

File: ops/docker-compose.yml (hunk 2 above)
Comment (annieke) [reply ops/docker-compose.yml:179]: talked to @tynes and will switch this to the singular dtl

File: integration-tests/sync-tests/sync-replica.spec.ts (hunk 3 above)
Comment (annieke) [reply integration-tests/sync-tests/sync-replica.spec.ts:33]: hmm i'm not sure but this does work 🤔 will investigate, @tynes is there a better log to grep for here?

File: ops/docker-compose.yml (hunk 2 above)
Comment (annieke) [reply ops/docker-compose.yml:179]: ^just tried using the existing DTL and crashed the whole stack when i started the integration test due to [this error](https://github.com/ethereum-optimism/optimism/pull/981/files#r641994439) in the comment above; then when i set the highestBlock to default to 0 i get this error instead

```
Cannot read property 'index' of undefined","stack":"TypeError: Cannot read property 'index' of undefined\n    at Object.parseBlock (/opt/optimism/packages/data-transport-layer/dist/src/services/l2-ingestion/handlers/transaction.js:10:46)\n    at L2IngestionService._syncSequencerBlocks (/opt/optimism/packages/data-transport-layer/dist/src/services/l2-ingestion/service.js:121:68)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async L2IngestionService._start (/opt/optimism/packages/data-transport-layer/dist/src/services/l2-ingestion/service.js:69:17)\n    at async L2IngestionService.start (/opt/optimism/packages/common-ts/dist/base-service.js:30:9)\n    at async Promise.all (index 2)\n    at async L1DataTransportService._start (/opt/optimism/packages/data-transport-layer/dist/src/services/main/service.js:48:9)\n    at async L1DataTransportService.start (/opt/optimism/packages/common-ts/dist/base-service.js:30:9)\n    at async /opt/optimism/packages/data-transport-layer/dist/src/services/run.js:61:9"},"msg":"Caught an unhandled error"
```
think the cause is we can't read the genesis block correctly in the dtl, which i could try fix in a followup PR but isn't within scope of this one. will keep this setup for now and chat with the team to see if there's a cleaner solution

File: ops/README.md (hunk 1 above)
Comment (annieke) [reply ops/README.md:43]: think we'll run into a genesis block error if we do that https://github.com/ethereum-optimism/optimism/pull/981#discussion_r643538928 might consider just changing the sleep condition to `>=` [here](https://github.com/ethereum-optimism/optimism/blob/245136f14afed2a731a5ca5f9b5b071a3131961b/packages/data-transport-layer/src/services/l2-ingestion/service.ts#L89)

File: ops/docker-compose.yml (hunk 2 above)
Comment (tynes) [reply ops/docker-compose.yml:179]: The DTL should never try to ingest the genesis block as it has no transactions

File: integration-tests/sync-tests/sync-replica.spec.ts (hunk 3 above)
Comment (tynes) [reply integration-tests/sync-tests/sync-replica.spec.ts:33]: You should be able to poll the RPC to do it in a generalized way. Note that if this works, then its not running as a verifier or you are getting logs from the wrong container because the log does say `Starting Verifier Loop`

https://github.com/ethereum-optimism/optimism/blob/86708bb5758cd2b647b3ca2be698beb5aa3af81f/l2geth/rollup/sync_service.go#L319

File: ops/docker-compose.yml (hunk 2 above)
Comment (tynes) [reply ops/docker-compose.yml:179]: This may be useful for deduplication of config: https://nickjanetakis.com/blog/docker-tip-82-using-yaml-anchors-and-x-properties-in-docker-compose

File: ops/README.md (hunk 1 above)
Comment (tynes) [reply ops/README.md:43]: This is definitely a bug in the DTL, we have had these sorts of problems in the past. A fix was added here, the `-1`

https://github.com/ethereum-optimism/optimism/blob/5e3c5d1c65ef3352947989ba9f51a1047bf7ba0c/packages/data-transport-layer/src/services/server/service.ts#L230

Not sure if that was the most ideal solution but it worked at the time, perhaps we can make another similar change to fix the problem?

Comment (annieke) [reply ops/README.md:43]: put in this fix: https://github.com/ethereum-optimism/optimism/pull/981/files#diff-f52a0d7bb7b1308354b3ee1e2099fba24225cceb77b8964638d239375ab10235R89-R93

File: integration-tests/sync-tests/sync-replica.spec.ts (hunk 3 above)
Comment (annieke) [reply integration-tests/sync-tests/sync-replica.spec.ts]: @tynes corrected the config and updated the log i grep for, but lmk if you think there's a better way to query the replica to ensure it's up!

File: ops/docker-compose.yml [hunk 4]
```diff
@@ -141,23 +142,47 @@ services:
...
         ROLLUP_STATE_DUMP_PATH: http://deployer:8081/state-dump.latest.json
-        # used for getting the addresses
         URL: http://deployer:8081/addresses.json
-        # connecting to the DTL
         ROLLUP_CLIENT_HTTP: http://dtl:7878
+        ROLLUP_BACKEND: 'l1'
         ETH1_CTC_DEPLOYMENT_HEIGHT: 8
         RETRIES: 60
-        IS_VERIFIER: "true"
+        ROLLUP_BACKEND: 'l1'
```
Comment (gakonst) [inline ops/docker-compose.yml:156]:
```suggestion
```

Duplicate

File: ops/docker-compose.yml [hunk 5]
```diff
@@ -62,6 +62,7 @@ services:
         # connect to the 2 layers
         DATA_TRANSPORT_LAYER__L1_RPC_ENDPOINT: http://l1_chain:8545
         DATA_TRANSPORT_LAYER__L2_RPC_ENDPOINT: http://l2geth:8545
+        DATA_TRANSPORT_LAYER__SYNC_FROM_L2: 'true'
```
Comment (gakonst) [inline ops/docker-compose.yml:65]: Why is this needed? I thought the DTL was able to be consumed by both replicas and verifiers?

**Review by gakonst — APPROVED** [review]

File: ops/docker-compose.yml (hunk 5 above)
Comment (annieke) [reply ops/docker-compose.yml:65]: without this switch the dtl wasn't syncing with L2 automatically

File: integration-tests/sync-tests/sync-replica.spec.ts (hunk 3 above)
Comment (tynes) [reply integration-tests/sync-tests/sync-replica.spec.ts]: I feel like polling the RPC would prevent an unknowning future person from removing/altering that logline in a way that it causes these tests to fail

File: integration-tests/sync-tests/2-sync-replica.spec.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  let env: OptimismEnv
+  let wallet: Wallet
+  let replica: DockerComposeNetwork
+  let provider: providers.JsonRpcProvider
+
+  const sequencerProvider = injectL2Context(l2Provider)
+
+  /* Helper functions */
+
+  const startReplica = async () => {
```
Comment (tynes) [inline integration-tests/sync-tests/2-sync-replica.spec.ts:20]: This function is basically the same as `startVerifier` - it should be possible to have a common file where both the verifier and replica can use that deduplicates the code

Comment (annieke) [reply integration-tests/sync-tests/2-sync-replica.spec.ts:20]: ^agreed, will refactor this!

Comment (annieke) [reply integration-tests/sync-tests/2-sync-replica.spec.ts:25]: @tynes i switched both to poll the endpoint instead of waiting for docker logs! as a long term vision when both verifier and replicas are stable, i think these tests can go for a larger refactor in which they share the same test suite of transactions and we just use a `verifier` or `replica` switch to run each. 

as of now though, the replica has matching state roots for pretty much everything but the verifier is the opposite, so i'm keeping them as separate files (also not in the `test` folder for now) so we could buff up the replica tests before uniswap launch.

Comment (tynes) [reply integration-tests/sync-tests/2-sync-replica.spec.ts:25]: Ok! Thats a good idea!

**Review by snario — APPROVED** [review]

**Review by tynes — APPROVED** [review]

File: ops/docker-compose.yml [hunk 7]
```diff
@@ -141,23 +142,46 @@ services:
...
+    env_file:
+      - ./envs/geth.env
+    environment:
+        ETH1_HTTP: http://l1_chain:8545
+        ROLLUP_STATE_DUMP_PATH: http://deployer:8081/state-dump.latest.json
+        URL: http://deployer:8081/addresses.json
+        ROLLUP_CLIENT_HTTP: http://dtl:7878
+        ROLLUP_BACKEND: 'l2'
+        ROLLUP_VERIFIER_ENABLE: 'true'
+        ETH1_CTC_DEPLOYMENT_HEIGHT: 8
```
Comment (snario) [inline ops/docker-compose.yml:180]: There doesn't seem to be any reason for us to be supplying this parameter. It just happens to be that the CTC is deployed at block 8 right now. Let's make an issue for this.

**Review by snario — APPROVED** [review]
